pkg/stub/handler.go

```diff
@@ -606,18 +638,66 @@ func (h *Handler) CleanUpOpenshiftNamespaceOnDelete(srcfg *v1alpha1.SamplesResou
 			err = h.templateclientwrapper.Delete("openshift", temp.Name, &metav1.DeleteOptions{})
 			if err != nil && !kerrors.IsNotFound(err) {
 				logrus.Warnf("Problem deleting openshift template %s on SamplesResource delete: %#v", temp.Name, err)
+			} else {
+				err = nil
 			}
 		}
 	}
 
 	err = h.secretclientwrapper.Delete("openshift", v1alpha1.SamplesRegistryCredentials, &metav1.DeleteOptions{})
 	if err != nil && !kerrors.IsNotFound(err) {
 		logrus.Warnf("Problem deleting openshift secret %s on SamplesResource delete: %#v", v1alpha1.SamplesRegistryCredentials, err)
+	} else {
+		err = nil
 	}
 
 	err = h.configmapclientwrapper.Delete(h.namespace, v1alpha1.SamplesResourceName)
 	if err != nil && !kerrors.IsNotFound(err) {
 		logrus.Warnf("Problem deleting openshift configmap %s on SamplesResource delete: %#v", v1alpha1.SamplesResourceName, err)
+	} else {
+		err = nil
+	}
+	return nil
+}
+
+// ProcessManagementField returns true if the operator should handle the SampleResource event
+// and false if it should not, as well as an err in case we want to bubble that up to
+// the controller level logic for retry
+func (h *Handler) ProcessManagementField(srcfg *v1alpha1.SamplesResource) (bool, error) {
+	switch srcfg.Spec.ManagementState {
+	case operatorsv1alpha1api.Removed:
+		if srcfg.ConditionFalse(v1alpha1.SamplesExist) {
```

Comment:
isn't it possible the credentials exist but the samples don't, in which case the credentials still need to be cleaned up?

Reply:
No either deleting the samplesresource or marking it removed will delete the credential created by the operator as well as the samples.

Reply:
I think we still need to get/confirm consensus here as well @bparees

Reply:
Sorry i'm not seeing it, what's the code flow that leads to the credential being deleted when the samplesresource is marked removed (particularly is samplesexist is false)?

Reply:
Flow I tested:
- default bringup, mgmt state is managed, samples installed, exist is true
- admin creats secret in operator's namespace
- operator creates secret in openshift namespace
- admin sets mgmt to removed
- first time through, exists is still true, calls `CleanUpOpenshiftNamespaceOnDelete`
- `CleanUpOpenshiftNamespaceOnDelete` deletes the samples and the secret in the openshift namespace, marks exist/cred conditions as false, updates samplesreource
- samples resource update reflecting updated conditions ... ignores since already processed given exists is false ... checking credentials is false is redundant

Only `CleanUpOpenshiftNamespaceOnDelete` removes things ... it always removes both samples and secret.

Reply:
right, but you could have a scenario where the secret exists but the samples do not exist, in which case this code is going to skip over the cleanup and not remove the secret.

Reply:
actually, no, I do not think you can have such a scenario now ... can you describe how you think we can get in such a state?

Reply:
1) SamplesResource is created, references rhel.  We do not create samples because we don't have a secret yet.

2) User creates their secret, we copy the secret into the openshift namespace and update the resource to samplesexist=true

3) User marks the resource to "Removed" either before we attempt to create the samples, or we encounter a failure trying to create the samples and then the user marks the resource "Removed".

Reply:
those are some tight timing windows but sure, will adjust if-check to account for them

Reply:
the case where we are failing to create the samples for some reason isn't necessarily a tight timing window, we could be in that state for a long time if something gets screwed up in RBAC or we have broken samples or something else.

it's also possible in the future we introduce new scenarios where we end up there (perhaps we introduce a second gating condition on creating the samples, such that it becomes easy to get in a scenario where you've created the secret, but can't create the samples yet because you're waiting for another condition still).  So I want to make sure we're taking all possible states into account.

Reply:
(I know you're convinced/acquiescing on this specific case, but in general I think those are the kinds of things we need to be considering when we are building our operator state machines)